README.md
# id3
A cross platform dart package to extract meta data from mp3 files.

This package contains functions that could extract meta tags from mp3
files that uses ``ID3 tag version 2.3.0`` and ``ID3 tag version 1`` to store meta data.

## Usage

```dart
import 'dart:io';
import 'package:id3/id3.dart';

void main(){
  List<int> mp3Bytes = File("./file.mp3").readAsBytesSync();
  MP3Instance mp3instance = new MP3Instance(mp3Bytes);

  /// parseTagsSync() returns 
  // 'true' if successfully parsed
  // 'false' if was unable to recognize tag so can't be parsed

  if(mp3instance.parseTagsSync()){
    print(mp3instance.getMetaTags());
  }
}

/// mp3instance.getMetaTags() returns Map<String, dynamic>
// {
//   "Title": "SongName",
//   "Artist": "ArtistName",
//   "Album": "AlbumName",
//   "APIC": {
//     "mime": "image/jpeg",
//     "textEncoding": "0",
//     "picType": "0",
//     "description": "description",
//     "base64": "AP/Y/+AAEEpGSUYAAQEBAE..."
//   }
// }
```

# Migrating from v0.1 to v1.0

- `MP3Instance` used to take filename as string in v0.1, which is updated to take bytes as `List<int>` in v1.0

```diff
- MP3Instance mp3instance = new MP3Instance("./file.mp3");
+ List<int> mp3Bytes = File("./file.mp3").readAsBytesSync();
+ MP3Instance mp3instance = new MP3Instance(mp3Bytes);
```

CHANGELOG.md
## [1.0.2] - 2021-11-08
* [Parse PicType, fix picture on ID3v2.2](https://github.com/sanket143/id3/pull/19)
* Welcome [@moffatman](https://github.com/moffatman) as a collaborator
* [Properly parse ID3v2.2 tags](https://github.com/sanket143/id3/pull/14)

## [1.0.1] - 2021-08-28
* [Properly parse ID3v2.2 tags](https://github.com/sanket143/id3/pull/14), with some explicit type notation

## [1.0.0] - 2021-07-17
* [API Update](https://github.com/sanket143/id3/pull/10), takes bytes instead of fileName as argument in `MP3Instance`
* [Fix corrupted album art](https://github.com/sanket143/id3/pull/9) [Thanks to [@moffotman](https://github.com/moffatman)]
* [Migration to null safety](https://github.com/sanket143/id3/pull/6) [Thanks to [@4-alok](https://github.com/4-alok)]

## [0.1.8] - 2020-01-08

* Removed redundant "this" and "new" keyword

## [0.1.3] - 2019-08-07

* Renamed parseTagsSync to parseTagsSync

## [0.1.1] - 2019-07-24

* First public release with id3v1 and id3v2.3 support

## [0.0.5] -  2019-07-19

* Added id3v1 support

## [0.0.4] -  2019-07-18

* Initial Release 


test/id3_test.dart
import 'package:id3/id3.dart';

void main() {

}


lib/src/const.dart
const Map<String, String> FRAMESv2_2 = {
  'CNT': 'PlayCounter',
  'COM': 'Comment',
  'CRA': 'Encryption',
  'EQU': 'Equalization',
  'IPL': 'People',
  'LNK': 'Information',
  'PIC': 'APIC',
  'POP': 'Popularimeter',
  'REV': 'Reverb',
  'SLT': 'Transcript',
  'TAL': 'Album',
  'TBP': 'BPM',
  'TCM': 'Composer',
  'TCO': 'ContentType',
  'TCR': 'Copyright',
  'TDA': 'Date',
  'TEN': 'EncodedBy',
  'TFT': 'FileType',
  'TIM': 'Time',
  'TKE': 'Key',
  'TLA': 'Language',
  'TLE': 'Length',
  'TMT': 'MediaType',
  'TOA': 'OriginalPerformer',
  'TOF': 'OriginalFilename',
  'TOL': 'OriginalLyricist',
  'TOR': 'OriginalReleaseYear',
  'TOT': 'OriginalAlbum',
  'TP1': 'Artist',
  'TP2': 'Accompaniment',
  'TP3': 'Conductor',
  'TP4': 'ModifiedBy',
  'TPB': 'Publisher',
  'TRC': 'ISRC',
  'TRK': 'Track',
  'TSI': 'Size',
  'TSS': 'Settings',
  'TT1': 'Content',
  'TT2': 'Title',
  'TT3': 'SubTitle',
  'TXT': 'Lyricist',
  'TXX': 'AdditionalInfo',
  'TYE': 'Year',
  'UFI': 'ID',
  'ULT': 'UnsyncTranscript',
  'WAF': 'OFileWebpage',
  'WAR': 'OArtistWebpage',
  'WAS': 'OAudioWebpage',
  'WCM': 'CommercialInfo',
  'WCP': 'CopyrightInfo',
  'WPB': 'OPublisherWebpage',
  'WXX': 'CLink'
};

const Map<String, String> FRAMESv2_3 = {
  'AENC': 'Encryption',
  'APIC': 'APIC',
  'COMM': 'Comment',
  'COMR': 'Commercial',
  'ENCR': 'ENCR',
  'EQUA': 'Equalization',
  'ETCO': 'ETCO',
  'GEOB': 'GEOB',
  'GRID': 'GRID',
  'IPLS': 'People',
  'LINK': 'Information',
  'MCDI': 'MCDI',
  'MLLT': 'MLLT',
  'OWNE': 'OWNE',
  'PRIV': 'PRIV',
  'PCNT': 'PlayCounter',
  'POPM': 'Popularimeter',
  'POSS': 'POSS',
  'RBUF': 'RBUF',
  'RVAD': 'RVAD',
  'SYLT': 'Lyrics',
  'SYTC': 'SYTC',
  'TALB': 'Album',
  'TBPM': 'BPM',
  'TCOM': 'Composer',
  'TCON': 'Genre',
  'TCOP': 'Copyright',
  'TDAT': 'Date',
  'TDLY': 'Playlist',
  'TENC': 'EncodedBy',
  'TEXT': 'Lyricist',
  'TFLT': 'FileType',
  'TIME': 'Time',
  'TIT1': 'Content',
  'TIT2': 'Title',
  'TIT3': 'SubTitle',
  'TKEY': 'Key',
  'TLAN': 'Language',
  'TLEN': 'Length',
  'TMED': 'MediaType',
  'TOAL': 'OriginalAlbum',
  'TOFN': 'OriginalFilename',
  'TOLY': 'OriginalLyricist',
  'TOPE': 'OriginalPerformer',
  'TORY': 'OriginalReleaseYear',
  'TOWN': 'Owner',
  'TPE1': 'Artist',
  'TPE2': 'Accompaniment',
  'TPE3': 'Conductor',
  'TPE4': 'ModifiedBy',
  'TPOS': 'TPOS',
  'TPUB': 'Publisher',
  'TRCK': 'Track',
  'TRDA': 'RecordedOn',
  'TRSN': 'RadioStation',
  'TRSO': 'RadioStationOwner',
  'TSIZ': 'Size',
  'TSRC': 'ISRC',
  'TSSE': 'Settings',
  'TYER': 'Year',
  'TXXX': 'AdditionalInfo',
  'UFID': 'ID',
  'USER': 'Terms',
  'USLT': 'UnsyncTranscript',
  'WCOM': 'CommercialInfo',
  'WCOP': 'CopyrightInfo',
  'WOAF': 'OFileWebpage',
  'WOAR': 'OArtistWebpage',
  'WOAS': 'OAudioWebpage',
  'WORS': 'ORadioWebpage',
  'WPAY': 'Payment',
  'WPUB': 'OPublisherWebpage',
  'WXXX': 'CLink'
};

const GENREv1 = [
  'Blues',
  'Classic Rock',
  'Country',
  'Dance',
  'Disco',
  'Funk',
  'Grunge',
  'Hip-Hop',
  'Jazz',
  'Metal',
  'New Age',
  'Oldies',
  'Other',
  'Pop',
  'Rhythm and Blues',
  'Rap',
  'Reggae',
  'Rock',
  'Techno',
  'Industrial',
  'Alternative',
  'Ska',
  'Death Metal',
  'Pranks',
  'Soundtrack',
  'Euro-Techno',
  'Ambient',
  'Trip-Hop',
  'Vocal',
  'Jazz & Funk',
  'Fusion',
  'Trance',
  'Classical',
  'Instrumental',
  'Acid',
  'House',
  'Game',
  'Sound clip',
  'Gospel',
  'Noise',
  'Alternative Rock',
  'Bass',
  'Soul',
  'Punk',
  'Space',
  'Meditative',
  'Instrumental Pop',
  'Instrumental Rock',
  'Ethnic',
  'Gothic',
  'Darkwave',
  'Techno-Industrial',
  'Electronic',
  'Pop-Folk',
  'Eurodance',
  'Dream',
  'Southern Rock',
  'Comedy',
  'Cult',
  'Gangsta',
  'Top 40',
  'Christian Rap',
  'Pop/Funk',
  'Jungle music',
  'Native US',
  'Cabaret',
  'New Wave',
  'Psychedelic',
  'Rave',
  'Showtunes',
  'Trailer',
  'Lo-Fi',
  'Tribal',
  'Acid Punk',
  'Acid Jazz',
  'Polka',
  'Retro',
  'Musical',
  'Rock ’n’ Roll',
  'Hard Rock'
];

const PICTYPE = {
  0x00: 'Other',
  0x01: 'FileIcon',
  0x02: 'OtherFileIcon',
  0x03: 'FrontCover',
  0x04: 'BackCover',
  0x05: 'LeafletPage',
  0x06: 'Media (e.g. lable side of CD)',
  0x07: 'LeadArtist',
  0x08: 'Artist',
  0x09: 'Conductor',
  0x0A: 'Band',
  0x0B: 'Composer',
  0x0C: 'Lyricist',
  0x0D: 'RecordingLocation',
  0x0E: 'DuringRecording',
  0x0F: 'DuringPerformance',
  0x10: 'VideoStill',
  // Might be a joke but it's in the ID3 spec
  0x11: 'ABrightColouredFish',
  0x12: 'Illustration',
  0x13: 'ArtistLogo',
  0x14: 'PublisherLogo'
};


lib/id3.dart
library id3;

import 'dart:convert';

import 'src/const.dart';

class MP3ParserException implements Exception {
  String cause;
  MP3ParserException(this.cause);
  String toString() {
    return this.cause;
  }
}

class _MP3FrameParser {
  List<int> buffer;
  int pos = 0;
  int lastEncoding = 0x00; // default to latin1
  _MP3FrameParser(this.buffer);
  List<int> readUntilTerminator(List<int> terminator,
      {bool aligned = false, bool terminatorMandatory = true}) {
    if (remainingBytes == 0) {
      return [];
    }
    for (int i = pos;
        i < buffer.length - (terminator.length - 1);
        i += (aligned ? terminator.length : 1)) {
      bool foundTerminator = true;
      for (int j = 0; j < terminator.length; j++) {
        if (buffer[i + j] != terminator[j]) {
          foundTerminator = false;
          break;
        }
      }
      if (foundTerminator) {
        final start = pos;
        pos = i + terminator.length;
        return buffer.sublist(start, pos - terminator.length);
      }
    }
    if (terminatorMandatory) {
      throw MP3ParserException(
          "Did not find terminator $terminator in ${buffer.sublist(pos)}");
    } else {
      return buffer.sublist(pos);
    }
  }

  String readLatin1String({bool terminator = true}) {
    return latin1
        .decode(readUntilTerminator([0x00], terminatorMandatory: terminator));
  }

  String readUTF16LEString({bool terminator = true}) {
    final bytes = readUntilTerminator([0x00, 0x00],
        aligned: true, terminatorMandatory: terminator);
    // final utf16les = List<int?>((bytes.length / 2).ceil());
    final utf16les = List.generate((bytes.length / 2).ceil(), (index) => 0);

    for (int i = 0; i < bytes.length; i++) {
      if (i % 2 == 0) {
        utf16les[i ~/ 2] = bytes[i];
      } else {
        utf16les[i ~/ 2] |= (bytes[i] << 8);
      }
    }
    return String.fromCharCodes(utf16les);
  }

  String readUTF16BEString({bool terminator = true}) {
    final bytes =
        readUntilTerminator([0x00, 0x00], terminatorMandatory: terminator);
    // final utf16bes = List<int?>((bytes.length / 2).ceil());
    final utf16bes = List.generate((bytes.length / 2).ceil(), (index) => 0);

    for (int i = 0; i < bytes.length; i++) {
      if (i % 2 == 0) {
        utf16bes[i ~/ 2] = (bytes[i] << 8);
      } else {
        utf16bes[i ~/ 2] |= bytes[i];
      }
    }
    return String.fromCharCodes(utf16bes);
  }

  String readUTF16String({bool terminator = true}) {
    final bom = buffer.sublist(pos, pos + 2);
    if (bom[0] == 0xFF && bom[1] == 0xFE) {
      pos += 2;
      return readUTF16LEString(terminator: terminator);
    } else if (bom[0] == 0xFE && bom[1] == 0xFF) {
      pos += 2;
      return readUTF16BEString(terminator: terminator);
    } else if (bom[0] == 0x00 && bom[1] == 0x00) {
      pos += 2;
      return "";
    } else {
      throw MP3ParserException(
          "Unknown UTF-16 BOM: $bom in ${buffer.sublist(pos)}");
    }
  }

  String readUTF8String({bool terminator = true}) {
    final bytes = readUntilTerminator([0x00], terminatorMandatory: terminator);
    return Utf8Decoder().convert(bytes);
  }

  void readEncoding() {
    if (buffer[pos] < 20) {
      if (lastEncoding == 0x01) {
        // Do not modify the BOM, 0x01 must apply to each field
        pos++;
      } else {
        lastEncoding = buffer[pos++];
      }
    }
  }

  String readString({bool terminator = true, bool checkEncoding: true}) {
    if (checkEncoding) {
      readEncoding();
    }
    if (pos == buffer.length) {
      return '';
    }
    if (lastEncoding == 0x00) {
      return readLatin1String(terminator: terminator);
    } else if (lastEncoding == 0x01) {
      return readUTF16String(terminator: terminator);
    } else if (lastEncoding == 0x02) {
      return readUTF16BEString(terminator: terminator);
    } else if (lastEncoding == 0x03) {
      return readUTF8String(terminator: terminator);
    } else {
      throw MP3ParserException(
          "Unknown Byte-Order Marker: $lastEncoding in $buffer");
    }
  }

  List<int> readBytes(int length) {
    pos += length;
    return buffer.sublist(pos - length, pos);
  }

  List<int> readRemainingBytes() {
    return buffer.sublist(pos);
  }

  int get remainingBytes {
    return buffer.length - pos;
  }
}

class MP3Instance {
  late final List<int> mp3Bytes;
  final Map<String, dynamic> metaTags = {};

  /// Member Functions
  MP3Instance(List<int> mp3Bytes) {
    this.mp3Bytes = mp3Bytes;
  }

  bool parseTagsSync() {
    List<int> _tag;
    _tag = mp3Bytes.sublist(0, 3);

    if (latin1.decode(_tag) == 'ID3') {
      final int major_v = mp3Bytes[3];
      final int revision_v = mp3Bytes[4];
      final int flag = mp3Bytes[5];

      final bool unsync = (0x40 & flag != 0);
      final bool extended = (0x20 & flag != 0);
      final bool experimental = (0x10 & flag != 0);

      metaTags['Version'] = 'v2.$major_v.$revision_v';

      if (extended) {
        print('Extended id3v2 tags are not supported yet!');
      } else if (unsync) {
        print('Unsync id3v2 tags are not supported yet!');
      } else if (experimental) {
        print('Experimental id3v2 tag');
      }

      int cb = 10;

      Map<String, String> frames_db = FRAMESv2_3;
      int frameNameLength = 4;
      int frameSizeLength = 4;
      int frameTagLength = 2;
      if (major_v == 2) {
        frames_db = FRAMESv2_2;
        frameNameLength = 3;
        frameSizeLength = 3;
        frameTagLength = 0;
      }
      final int frameHeaderLength =
          frameNameLength + frameSizeLength + frameTagLength;

      while (true) {
        final List<int> frameHeader =
            mp3Bytes.sublist(cb, cb + frameHeaderLength);
        final List<int> frameName = frameHeader.sublist(0, frameNameLength);

        final RegExp exp = RegExp(r'[A-Z0-9]+');
        if (latin1.decode(frameName) !=
            exp.stringMatch(latin1.decode(frameName))) {
          break;
        }

        final int frameSize = parseSize(
            frameHeader.sublist(
                frameNameLength, frameNameLength + frameSizeLength),
            major_v);
        final List<int> frameContent = mp3Bytes.sublist(
            cb + frameHeaderLength, cb + frameHeaderLength + frameSize);

        if (frames_db[latin1.decode(frameName)] == FRAMESv2_3['APIC']) {
          final Map<String, String> apic = {
            'mime': '',
            'textEncoding': frameContent[0].toString(),
            'picType': '',
            'description': '',
            'base64': ''
          };

          final frame = _MP3FrameParser(frameContent);
          frame.readEncoding();
          // In ID3v2.2 the MIME type is always 3 characters
          if (major_v == 2) {
            apic['mime'] = latin1.decode(frame.readBytes(3));
          } else {
            apic['mime'] = frame.readLatin1String();
          }
          apic['picType'] = PICTYPE[frame.readBytes(1).first] ?? 'Unknown';
          apic['description'] = frame.readString(checkEncoding: false);
          apic['base64'] = base64.encode(frame.readRemainingBytes());
          metaTags['APIC'] = apic;
        } else if (frames_db[latin1.decode(frameName)] == FRAMESv2_3['USLT']) {
          final frame = _MP3FrameParser(frameContent);
          frame.readEncoding();
          final language = latin1.decode(frame.readBytes(3));
          String contentDescriptor;
          contentDescriptor = frame.readString(checkEncoding: false);
          final lyrics = (frame.remainingBytes > 0)
              ? frame.readString(checkEncoding: false, terminator: false)
              : contentDescriptor;
          if (frame.remainingBytes == 0) {
            contentDescriptor = '';
          }
          metaTags['USLT'] = {
            'language': language,
            'contentDescriptor': contentDescriptor,
            'lyrics': lyrics
          };
        } else if (frames_db[latin1.decode(frameName)] == FRAMESv2_3['WXXX']) {
          final frame = _MP3FrameParser(frameContent);
          metaTags['WXXX'] = {
            'description': frame.readString(),
            'url': frame.readLatin1String(terminator: false)
          };
        } else if (frames_db[latin1.decode(frameName)] == FRAMESv2_3['COMM']) {
          final frame = _MP3FrameParser(frameContent);
          frame.readEncoding();
          final language = latin1.decode(frame.readBytes(3));
          final shortDescription = frame.readString(checkEncoding: false);
          final text =
              frame.readString(terminator: false, checkEncoding: false);
          if (metaTags['COMM'] == null) {
            metaTags['COMM'] = {};
            if (metaTags['COMM'][language] == null) {
              metaTags['COMM'][language] = {};
            }
          }
          metaTags['COMM'][language][shortDescription] = text;
        } else if (frames_db[latin1.decode(frameName)] == FRAMESv2_3['MCDI'] ||
            frames_db[latin1.decode(frameName)] == FRAMESv2_3['RVAD']) {
          // Binary data
          metaTags[frames_db[latin1.decode(frameName)] ??
              latin1.decode(frameName)] = frameContent;
        } else {
          final String tag =
              frames_db[latin1.decode(frameName)] ?? latin1.decode(frameName);
          metaTags[tag] =
              _MP3FrameParser(frameContent).readString(terminator: false);
        }

        cb += frameHeaderLength + frameSize;
      }

      return true;
    }

    final List<int> _header =
        mp3Bytes.sublist(mp3Bytes.length - 128, mp3Bytes.length);
    _tag = _header.sublist(0, 3);

    if (latin1.decode(_tag).toLowerCase() == 'tag') {
      metaTags['Version'] = '1.0';

      final List<int> _title = _header.sublist(3, 33);
      final List<int> _artist = _header.sublist(33, 63);
      final List<int> _album = _header.sublist(63, 93);
      final List<int> _year = _header.sublist(93, 97);
      final List<int> _comment = _header.sublist(97, 127);
      final int _genre = _header[127];

      metaTags['Title'] = latin1.decode(_title).trim();
      metaTags['Artist'] = latin1.decode(_artist).trim();
      metaTags['Album'] = latin1.decode(_album).trim();
      metaTags['Year'] = latin1.decode(_year).trim();
      metaTags['Comment'] = latin1.decode(_comment).trim();
      metaTags['Genre'] = GENREv1[_genre];

      return true;
    }

    return false;
  }

  Map<String, dynamic>? getMetaTags() {
    return metaTags;
  }
}

int parseSize(List<int> block, int major_v) {
  int len;
  if (major_v == 4) {
    assert(block.length == 4);
    len = block[0] << 21;
    len += block[1] << 14;
    len += block[2] << 7;
    len += block[3];
  } else if (major_v == 3) {
    assert(block.length == 4);
    len = block[0] << 24;
    len += block[1] << 16;
    len += block[2] << 8;
    len += block[3];
  } else if (major_v == 2) {
    assert(block.length == 3);
    len = block[0] << 16;
    len += block[1] << 8;
    len += block[2];
  } else {
    throw MP3ParserException("Unknown major version $major_v");
  }

  return len;
}

List<int> cleanFrame(List<int> bytes) {
  List<int> temp = new List<int>.from(bytes);

  temp.removeWhere((item) => item < 1);

  if (temp.length > 3) {
    return temp.sublist(3);
  } else {
    return temp;
  }
}

List<int> removeZeros(List<int> bytes) {
  return bytes.where((i) => i != 0).toList();
}
